pkg/authz/permissions.go

```diff
@@ -79,3 +90,100 @@ func checkAccess(ctx context.Context, c *kubernetes.Clientset, np NamespacedPerm
 
 	return r.Status.Allowed, nil
 }
+
+func createUnwantedPermissions(resourceRules []authzv1.ResourceRule, ns string, definedRules map[[4]string]int) []string {
+	unwantedPermissions := []string{}
+	for _, rr := range resourceRules {
+		APIGroups := rr.APIGroups
+		if len(APIGroups) == 0 {
+			APIGroups = []string{""}
+		}
+		ResourceNames := rr.ResourceNames
+		if len(ResourceNames) == 0 {
+			ResourceNames = []string{""}
+		}
+		for _, ag := range APIGroups {
+			for _, rs := range rr.Resources {
+				for _, rn := range ResourceNames {
+					for _, vb := range rr.Verbs {
+						if ag == "*" || rs == "*" {
+							unwantedPermissions = append(
+								unwantedPermissions,
+								fmt.Sprintf(
+									"Excess permission: namespace=%s,resource=%s,resource-name=%s,verbs=%s",
+									ns, strings.Join([]string{rs, ag}, "."), rn, vb,
+								),
+							)
+						}
+						if _, ok := definedRules[[4]string{ag, rs, rn, vb}]; !ok {
+							unwantedPermissions = append(
+								unwantedPermissions,
+								fmt.Sprintf(
+									"Excess permission: namespace=%s,resource=%s,resource-name=%s,verbs=%s",
+									ns, strings.Join([]string{rs, ag}, "."), rn, vb,
+								),
+							)
+						}
+					}
+				}
+			}
+		}
+	}
+	return unwantedPermissions
+}
+func createUnwantedPermissionsList(ctx context.Context, cfg *rest.Config, ns string, definedRules map[[4]string]int) ([]string, error) {
+	c, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return []string{}, fmt.Errorf("error creating the client: %w", err)
+	}
+
+	ssrr := authzv1.SelfSubjectRulesReview{
+		ObjectMeta: metav1.ObjectMeta{Namespace: ns},
+		Spec:       authzv1.SelfSubjectRulesReviewSpec{Namespace: ns},
+	}
+
+	r, err := c.AuthorizationV1().SelfSubjectRulesReviews().Create(ctx, &ssrr, metav1.CreateOptions{})
+	if err != nil {
+		return []string{}, err
+	}
+	resourceRules := r.Status.ResourceRules
+
+	unwantedPermissions := createUnwantedPermissions(resourceRules, ns, definedRules)
+
+	return unwantedPermissions, nil
+}
+
+func AccessList(ctx context.Context, cfg *rest.Config, namespaces []string, pl []Permission) ([]string, error) {
+	unwantedPermissions := []string{}
+	for _, ns := range namespaces {
+
+		definedRules := make(map[[4]string]int)
+		for _, p := range pl {
+			APIGroups := p.APIGroups
+			if len(APIGroups) == 0 {
+				APIGroups = []string{""}
+			}
+			ResourceNames := p.ResourceNames
+			if len(ResourceNames) == 0 {
+				ResourceNames = []string{""}
+			}
+			for _, ag := range APIGroups {
+				for _, rs := range p.Resources {
+					for _, rn := range ResourceNames {
+						for _, vb := range p.Verbs {
+							definedRules[[4]string{ag, rs, rn, vb}] = 1
+						}
+					}
+				}
+			}
```

Comment:
can we just refactor as follows:

```suggestion
			if len(p.APIGroups) == 0 || len(p.ResourceNames) == 0 {
				continue
			}
			for _, ag := range p.APIGroups {
				for _, rs := range p.Resources {
					for _, rn := range p.ResourceNames {
						for _, vb := range p.Verbs {
							definedRules[[4]string{ag, rs, rn, vb}] = 1
						}
					}
				}
			}
```

Reply:
We must iterate once when the length is zero with an empty string as a value. Continuing the loop will skip some of the rules.

Reply:
ok